File: 0120-triangle/0120-triangle.js
/**
 * @param {number[][]} triangle
 * @return {number}
 */
var minimumTotal = function(t, y=0, x=0, mem={}) {
    // last possible y 
    const ly = t.length-1
    // out of bound
    if(y > ly) return Number.MAX_VALUE
    // if y is inbound, check x for current y
    const lx = t[y].length-1    
    if(x > lx) return Number.MAX_VALUE

    // if bottom reached, return that value
    if(y===ly) return t[y][x]

    // all other cases
    const key = `${y}_${x}`
    if(!(key in mem)) {
        // take the value of current cell +
        // Min of(go down, or go down right)
        mem[key] = t[y][x] + Math.min( minimumTotal(t, y+1, x, mem), minimumTotal(t, y+1, x+1, mem) )
    }

    return mem[key]
};

File: 0120-triangle/0120-triangle.py
class Solution:
    # LESSON: do not initialize a dict on any object 
    # in python as default params, as they gets reused 
    # with all the function calls, even on repeated test cases
    def minimumTotal(self, triangle: List[List[int]], y=0, x=0, m=None) -> int:
        if m is None:
            m={}
        
        # if index out of bound
        # Bad Path
        if y > (len(triangle) - 1) or x > (len(triangle[y]) - 1):
            return float('inf')

        # last row reached
        if y == (len(triangle) - 1):
            return triangle[y][x]
        
        # use memoization
        if (y,x) not in m:
            m[(y,x)] = triangle[y][x] + min(self.minimumTotal(triangle, y+1, x, m), self.minimumTotal(triangle, y+1, x+1, m))

        return m[(y,x)]

        